### PR TITLE
Extend the help message when rustup acts as a proxy

### PR DESCRIPTION
**Pull request by ssomers**
Alternate but not quite right proposal for #2285

### CODE REVIEW

**Comment by ssomers** [comment]
I forgot to mention: `cargo` and `rustc` have the same [`ArgRequiredElseHelp`](https://docs.rs/clap/2.33.0/clap/enum.AppSettings.html#variant.ArgRequiredElseHelp), so the empty command line checks out. But `rustc` does not have subcommands, so `rustc help` looks crappy:
```
error: couldn't read help: The system cannot find the file specified. (os error 2)

error: aborting due to previous error


This proxy of rustc supports an additional +toolchain option, see `rustup help`.
```

In addition, I wonder if the -h and --help option exist like that on all platforms. It might have been /h in Windows (but it isn't) and Mac is uncharted territory for me.

File: src/cli/proxy_mode.rs [hunk 1]
```diff
@@ -38,12 +38,22 @@ pub fn main() -> Result<()> {
         } else {
             env::args_os().skip(2).collect()
         };
+        let help = cmd_args.is_empty()
+            || cmd_args == &["help"]
```
Comment (kinnison) [inline src/cli/proxy_mode.rs]: Do you want this to happen if someone does `cargo help build` ?  Maybe, maybe not.

File: src/cli/proxy_mode.rs [hunk 2]
```diff
@@ -38,12 +38,22 @@ pub fn main() -> Result<()> {
...
 
         let cfg = set_globals(false, true)?;
         cfg.check_metadata_version()?;
-        direct_proxy(&cfg, &arg0, toolchain, &cmd_args)?
+        (help, direct_proxy(&cfg, &arg0, toolchain, &cmd_args)?)
     };
 
+    if help {
+        println!(
+            "\nThis proxy of {} supports an additional +toolchain option, see `rustup help`.",
```
Comment (kinnison) [inline src/cli/proxy_mode.rs]: My suggestion of `rustup proxy` (perhaps `rustup proxy list`) would be a nicer spot for this to reference.  Of course, that doesn't exist yet...

**Review by kinnison — COMMENTED** [review]
I think this is getting somewhere.  It probably needs some tests to make sure it will work.

File: src/cli/proxy_mode.rs (hunk 1 above)
Comment (ssomers) [reply src/cli/proxy_mode.rs]: I couldn't have worded it better...

But now I think it shouldn't even hook onto `cargo help`. Once you dive into subcommands, you're on your own. And of course it avoids the awkward output of `rustc help`.

Comment (ssomers) [reply src/cli/proxy_mode.rs]: And by the way, there's no reason to expect extra flags besides -h or --help. `cargo -V -h` does not give you both version and help, and `cargo --color=always -h` does not colour your day. Help overrules almost everything, including  `cargo -h --baloney`, but not `cargo --baloney -h`.

But that still means it's better to compare the first argument and allow more arguments.

**Comment by ssomers** [comment]
Struggled to test stuff, and not sure exactly what I'm testing. I did't see clear in the naming of modules and test cases, so just added to the "misc" pile.

**Comment by ssomers** [comment]
Oops, never mind. This `direct_proxy` on non-Windows does an exec. I would suggest to put the Windows-only `process::exit` inside `run_command_for_dir` and not pretend that it returns.

We can't do anything after the `direct_proxy` call, and we can't just print some help text before it because the exec may fail (like when the proxy's target doesn't exist). To make this work, we're talking changing process treatment, at least for commands requesting help, and that's still emulating clap. It doesn't seem worth while.

**Comment by kinnison** [comment]
Aah I hadn't thought of that, sorry that you feel your time was wasted.  I still think there's potentially value in working out how to improve matters for users though.

**Comment by ssomers** [comment]
Sure, but I don't see any low hanging fruit any more. Unless you'd want to give up the unix-specific exec and do it with process::status or process::spawn, and possibly face some change in the way file descriptors or what not turn out.

It's not relevant on the road to better help, but I wonder if the current process::status isn't a (minor or philosophical) problem now: you start some cargo build on Windows, inadvertently keeping a lock on the rustup executable, and while that build is brewing, try to update rustup itself? Not that I ever saw `rustup self update` actually do anything. With a spawn instead, rustup could immediately exit in most cases, or wait on the child and complete the help message otherwise.

**Comment by ssomers** [comment]
I realised something else smelled fishy: I had to check the exitcode to avoid producing the extra help text when the proxied help couldn't appear because the proxied binary is missing. That doesn't make sense in hindsight because, on both unix and windows, `run_command_for_dir` returns no exitcode but `Err` if the executable is missing. Also, I'm reminded now that without the check on exitcode, the extra help message is produced 4 times (says the test framework). One hour later I still can't figure out what is actually happening.

**Pull request closed without merging**